src/geometry/algebra/lie_group.lean

```diff
@@ -0,0 +1,258 @@
+                        corners `I`.
+
+## Implementation notes
+A priori, a Lie group here is a manifold with corner.
+-/
+
+section lie_group
+
+set_option default_priority 100
+
+/-- A Lie (additive) group is a group and a smooth manifold at the same time in which
+the addition and negation operations are smooth. -/
+class lie_add_group {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+  {H : Type*} [topological_space H]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] (I : model_with_corners 𝕜 E H)
+  (G : Type*) [add_group G] [topological_space G] [topological_add_group G] [charted_space H G]
+  extends smooth_manifold_with_corners I G : Prop :=
+(smooth_add : smooth (I.prod I) I (λ p : G×G, p.1 + p.2))
+(smooth_neg : smooth I I (λ a:G, -a))
+
+/-- A Lie group is a group and a smooth manifold at the same time in which
+the multiplication and inverse operations are smooth. -/
+@[to_additive lie_add_group]
+class lie_group {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+  {H : Type*} [topological_space H]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] (I : model_with_corners 𝕜 E H)
+  (G : Type*) [group G] [topological_space G] [topological_group G] [charted_space H G]
+  extends smooth_manifold_with_corners I G : Prop :=
+(smooth_mul : smooth (I.prod I) I (λ p : G×G, p.1 * p.2))
+(smooth_inv : smooth I I (λ a:G, a⁻¹))
+
+section lie_group
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{H : Type*} [topological_space H]
+{E : Type*} [normed_group E] [normed_space 𝕜 E] {I : model_with_corners 𝕜 E H}
+{F : Type*} [normed_group F] [normed_space 𝕜 F] {J : model_with_corners 𝕜 F F}
+{G : Type*} [topological_space G] [charted_space H G] [group G]
+[topological_group G] [lie_group I G]
+{E' : Type*} [normed_group E'] [normed_space 𝕜 E']
+{H' : Type*} [topological_space H'] {I' : model_with_corners 𝕜 E' H'}
+{M : Type*} [topological_space M] [charted_space H' M] [smooth_manifold_with_corners I' M]
+{E'' : Type*} [normed_group E''] [normed_space 𝕜 E'']
+{H'' : Type*} [topological_space H''] {I'' : model_with_corners 𝕜 E'' H''}
+{M' : Type*} [topological_space M'] [charted_space H'' M'] [smooth_manifold_with_corners I'' M']
+
+@[to_additive]
+lemma smooth_mul : smooth (I.prod I) I (λ p : G×G, p.1 * p.2) :=
+lie_group.smooth_mul
+
+@[to_additive]
+lemma smooth.mul {f : M → G} {g : M → G} (hf : smooth I' I f) (hg : smooth I' I g) :
+  smooth I' I (f * g) :=
+smooth_mul.comp (hf.prod_mk hg)
+
+namespace lie_group
+
+/-- `L g` denotes left multiplication by `g` -/
+@[nolint unused_arguments, to_additive L_add "`L_add g` denotes left addition by `g`"]
+def L : G → G → G := λ g : G, λ x : G, g * x
+
+attribute [nolint unused_arguments] lie_add_group.L_add
+
+/-- `R g` denotes right multiplication by `g` -/
+@[nolint unused_arguments, to_additive R_add "`R_add g` denotes right addition by `g`"]
+def R : G → G → G := λ g : G, λ x : G, x * g
+
+attribute [nolint unused_arguments] lie_add_group.R_add
```

Comment:
This should absolutely go in another file (probably the earliest file in the library you can find on groups). There, you can define this (under the names `left_mul` and `left_add`) just under the assumptions `has_mul` and `has_add`. And introduce a localized notation in the Lie group file with `L` and `R`.